src/generators.ts

```diff
@@ -43,3 +32,54 @@ export function last<T>(iter: Iterable<T>): T {
     }
     return val;
 }
+
+export class MemoizedIterable<T> {
+    private iterator: Iterator<T>;
+    private nextIndex: number;
+    private complete: boolean;
+
+    public constructor(iter: Iterator<T> | Iterable<T>) {
+        // case when iter is Iterable
+        if (iter[Symbol.iterator]) {
+            this.iterator = iter[Symbol.iterator]();
+        } else {
+            this.iterator = iter as Iterator<T>;
+        }
+        this.nextIndex = 0;
+        this.complete = false;
+    }
+
+    public get length(): number {
+        while (!this.complete) {
+            this.nextValue();
+        }
+        return this.nextIndex;
+    }
+
+    private nextValue(): T {
+        const result = this.iterator.next();
+        if (result.done) {
+            this.complete = true;
+            return null;
+        }
+
+        this[this.nextIndex++] = result.value;
+        return result.value;
+    }
+
+    public *[Symbol.iterator](): Iterator<T> {
+        let index = 0;
+
+        while (index < this.nextIndex || !this.complete) {
```

Comment:
I think we can reduce the testing branching a little.

```js
        while (!this.complete) {
            // checks if there is a memoize
            while (index < this.nextIndex) {
                yield this[index++];
            }

            // this fills the next value
            this.nextValue();
        }
```

The left hand side of `||` in the first while isn't needed as the right hand will actually determine when the first while is done. It also avoid the first iteration for the first sub iterator where index and nextIndex are 0. In that first `this.complete` would determine whether to try or not.